src/v2/providers/database.ts

```diff
@@ -0,0 +1,452 @@
+ *
+ * @param referenceOrOpts - Options or a string reference.
+ * @param handler - Event handler which is run every time a Realtime Database deletion occurs.
+ */
+export function onRefDeleted(
+  referenceOrOpts: string | ReferenceOptions,
+  handler: (event: DatabaseEvent<DataSnapshot>) => any | Promise<any>
+): CloudFunction<DatabaseEvent<DataSnapshot>> {
+  return onOperation(
+    deletedEventType,
+    referenceOrOpts,
+    handler,
+    false
+  ) as CloudFunction<DatabaseEvent<DataSnapshot>>;
+}
+
+/** @internal */
+export function getOpts(referenceOrOpts: string | ReferenceOptions) {
+  let path: string, instance: string, opts: options.EventHandlerOptions;
+  if (typeof referenceOrOpts === 'string') {
+    path = normalizePath(referenceOrOpts);
+    instance = '*';
+    opts = {};
+  } else {
+    path = normalizePath(referenceOrOpts.ref);
+    instance = referenceOrOpts.instance || '*';
+    opts = { ...referenceOrOpts };
+    delete (opts as any).ref;
+    delete (opts as any).instance;
+  }
+
+  return {
+    path,
+    instance,
+    opts,
+  };
+}
+
+/** @hidden */
+function trimParam(param: string) {
+  const paramNoBraces = param.slice(1, -1);
+  if (paramNoBraces.includes('=')) {
+    return paramNoBraces.slice(0, paramNoBraces.indexOf('='));
+  }
+  return paramNoBraces;
+}
+
+/** @internal */
+export function matchParams(
+  pathSegments: string[],
+  wildcards: RegExpMatchArray,
+  refSegments: string[],
+  params: Record<string, string>
+): void {
+  let wildcardNdx = 0,
+    refNdx = 0;
+  for (
+    let pathNdx = 0;
+    pathNdx < pathSegments.length && wildcardNdx < wildcards.length;
+    pathNdx++
+  ) {
+    const multiSegmentWildcard = pathSegments[pathNdx].includes('**');
+
+    const remainingPathSegments = pathSegments.length - 1 - pathNdx;
+    const nextRefElement = refSegments.length - remainingPathSegments;
+
+    if (pathSegments[pathNdx] === wildcards[wildcardNdx]) {
+      params[trimParam(wildcards[wildcardNdx++])] = multiSegmentWildcard
+        ? refSegments.slice(refNdx, nextRefElement).join('/')
+        : refSegments[refNdx];
+    }
+
+    refNdx = multiSegmentWildcard ? nextRefElement : refNdx + 1;
+  }
+}
+
+/** @internal */
+export function makeParams(
+  event: RawRTDBCloudEvent,
+  path: string,
+  instance: string
+): Record<string, string> {
+  const params: Record<string, string> = {};
+
+  const pathWildcards = path.match(WILDCARD_REGEX);
+  if (pathWildcards) {
+    matchParams(path.split('/'), pathWildcards, event.ref.split('/'), params);
+  }
+
+  const instanceWildcards = instance.match(WILDCARD_REGEX);
+  if (instanceWildcards) {
+    matchParams([instance], instanceWildcards, [event.instance], params);
+  }
+
+  return params;
+}
+
+/** @hidden */
+function makeDatabaseEvent(
+  event: RawRTDBCloudEvent,
+  instance: string,
+  params: Record<string, string>
+): DatabaseEvent<DataSnapshot> {
+  const snapshot = new DataSnapshot(
+    event.data.delta,
```

Comment:
It's delta in both create and delete events? I assumed it would be `event.data.delta` for beforeRefCreated and `event.data.data` for beforeRefDeleted.

Reply:
Ahhh 🤦‍♂️ you're right